Add altmetric support for Handles and URIs

mghughes opened the pull request:
**JIRA Ticket**: [ISLANDORA-2374](https://jira.duraspace.org/browse/ISLANDORA-2374)

# What does this Pull Request do?

The altmetrics module currently only supports DOI identifiers when creating altmetric badges. The altmetric API has support for many other identifier types; this PR adds support for handles and URIs. By default, DOIs will still take precendence, but this is configurable in the admin.

# What's new?

* Add a draggable table to badges admin form for identifier weight/xpath configuration.
* Add new variables for handle xpath, doi xpath, and weights.
* Changes to altmetrics badge block to loop through the identifier types.

# How should this be tested?

Testing basic behaviour:
  * Add altmetrics block to somewhere visible
  * In badges configuration:
    * Add cmodel you'll be testing with to the "display badges list"
  * Ingest object that will display a badge.
  * Replace MODS datastreams with the following examples containing DOI, handle and URI respectively.
      https://gist.github.com/mghughes/1ff31ef4d85344354b8e204703db0227
      https://gist.github.com/mghughes/e5c1c4c9e674692fd9554a5808246af8
      https://gist.github.com/mghughes/9f354c88917b04c37c7b6b196dce1b4a
  * Ensure badge displays and link corresponds to identifier in MODS.

Testing ordering:
  * In badges admin coniguration, play around with ordering of identifier types.
  * Ingest an object that will display a badge.
  * Replace MODS datastream with the following (contains all 3 identifier types)
      https://gist.github.com/mghughes/edd5c502a47efb559d9a4cdce45e479e
  * Ensure badge displays and corresponds to first identifier
  * Repeat, changing order and which identifiers present in MODS.

TESTING XPaths:
  Xpath retrieval code is the same as before, but for completeness it wouldn't hurt to test this
  * In badges admin configuration, change the values of the xpath text fields
  * Change identifier locations in sample MODS file to correspond to the values given
  * Ingest object, replace MODS with altered sample MODS
  * Ensure badge displays and links correctly

# Additional Notes:
Variables have been added individually for each identifier xpath and weight, which seemed most straightforward in order to not break existing install which use the "doi_xpath" variable. But it may be worth considering just saving a blob instead? Open to ideas here.

# Interested parties
@bondjimbond

Commit:
Admin changes for adding new identifier support

Replaces the doi xpath textfield with a draggable table. Adds handle
and uri identifier options.

XPath and weight are stored in a separate variable for each identifier
type, which seems to me the simplest way to keep this compatible with existing
implementations using the "islandora_badges_doi_xpath" variable.

Commit:
Add handle and uri support to altmetrics

bondjimbond commented:
Thanks for this PR! This is a long-desired feature that I've just never had the time to dig into. I'll check it out today.

jordandukart commented:
@bondjimbond will be giving this a once-over as part of our internal review process here as well.

jordandukart commented on css/islandora_badges.admin.css:
```diff
@@ -0,0 +1,27 @@
+/* Admin style sheet for islandora badges */
+
+/* drag handler placeholder cell */
+.islandora-solr-multiple-drag
```
Copypasta.

jordandukart commented on includes/admin.form.inc:
```diff
@@ -5,23 +5,107 @@
+    '#theme' => 'islandora_badges_admin_identifier_fields',
   );
 
+  $terms = [];
```
PHP 5.4 specific unfortunately: http://php.net/manual/en/language.types.array.php

mghughes replied:
Ha, that's python coming back to bite me I think

bondjimbond commented on includes/admin.form.inc:
```diff
@@ -5,23 +5,107 @@
+  $form['islandora_badges_identifier_table'] = array(
+    '#title' => t('Identifier Fields'),
+    '#type' => 'item',
+    '#description' => t('MODS XPaths for retrieving identifier information. Set weights to change the order of identifiers to look for. DOI is first by default.'),
```
It's important to keep the note that XPATH "type" attribute is case-sensitive.

bondjimbond commented:
Why just handle.net and URIs? Altmetric supports a whole range of identifiers: https://help.altmetric.com/support/solutions/articles/6000134562-what-scholarly-identifiers-are-supported-by-altmetric-

If we're going to expand, the best approach in the long term (and forestalling another eight pull requests like this one) would be to support all of them. Although of course such a long list of XPATHs would be rather onerous... 

If we were to go there, a Solr field that can be parsed to grab the "type" attribute and sort it out accordingly (e.g. https://github.com/bondjimbond/islandora_rightsstatements/blob/7.x/includes/utilities.inc#L82-L113) would be a good way to do it... if solr field contains "handle" then it's a handle, if contains "isbn" then it's an isbn, etc.

Would such a change be too far outside the scope of this PR?

jordandukart commented on modules/islandora_altmetrics/islandora_altmetrics.module:
```diff
@@ -47,24 +47,28 @@ function islandora_altmetrics_block_view($delta) {
+          $id_value = islandora_badges_get_identifier_value($object, $id_type, $id_xpath);
+          if ($id_value) {
+            // Only generate block if a result exists - check API callback.
+            $file_headers = @get_headers("http://api.altmetric.com/v1/" . $id_type . "/" . $id_value . "?callback=my_callback");
```
Know this was already here but we really should look at using [drupal_http_request](https://api.drupal.org/api/drupal/includes%21common.inc/function/drupal_http_request/7.x) while passing the method as `HEAD`. The format it returns is nicer to check against for a status code imo. Similarly, the existing `get_headers` call isn't actually doing a HEAD request.

Commit:
Fixup silly things

Commit:
Add note about xpath being case-sensitive

Commit:
Use drupal_http_request instead of get_headers

mghughes commented:
@bondjimbond I do agree that dynamically building the “types” with something like a solr field is the way to go, but that does seem out of scope here and requires a pretty significant effort. For now we’ve just gone with handle.net and URI since they seem the most prevalent, and therefore most immediately useful.

jordandukart commented on modules/islandora_altmetrics/islandora_altmetrics.module:
```diff
@@ -52,8 +52,9 @@ function islandora_altmetrics_block_view($delta) {
-            $file_headers = @get_headers("http://api.altmetric.com/v1/" . $id_type . "/" . $id_value . "?callback=my_callback");
-            if ($file_headers['0'] != 'HTTP/1.1 404 Not Found') {
+            $altmetrics_api_url = 'http://api.altmetric.com/v1/' . $id_type . '/' . $id_value . '?callback=my_callback';
+            $head_response = @drupal_http_request($altmetrics_api_url, array('method' => 'HEAD'));
```
Shouldn't need the `@` operator anymore.

jordandukart commented on includes/admin.form.inc:
```diff
@@ -5,23 +5,107 @@
+  $form['islandora_badges_identifier_table'] = array(
+    '#title' => t('Identifier Fields'),
+    '#type' => 'item',
+    '#description' => t('MODS XPaths for retrieving identifier information. Set weights to change the order of identifiers to look for. DOI is first by default. Note that Xpaths are case-sensitive.'),
```
`Xpath`?

Commit:
Fix capitalization

Commit:
Remove @ operator

bondjimbond commented:
It seems to work quite well. @Islandora/7-x-1-x-committers any comments/thoughts on the approach?

Commit:
Only show block if 200 response

Other than 200 and 404, 403, 429, 502 are also possible according
to the altmetric API docs. So we don't want to render the block
if those are returned either.

BrandonMorr commented:
This has passed the DGI QA test gauntlet 👍

bondjimbond commented on includes/admin.form.inc:
```diff
@@ -5,23 +5,107 @@
  * Admin form functions for Islandora Altmetrics.
```
Not exactly part of the PR, but would be appreciated if you could throw in a correction to this error in the documentation (should be Islandora Badges, no Islandora Altmetrics)

bondjimbond commented on includes/admin.form.inc:
```diff
@@ -5,23 +5,107 @@
  * Admin form functions for Islandora Altmetrics.
```
Not exactly part of the PR, but would be appreciated if you could throw in a correction to this error in the documentation (should be Islandora Badges, no Islandora Altmetrics)

mghughes replied:
No problem - done now.

Commit:
Fix module name

mghughes commented:
@bondjimbond Happy with this one? Just wondering if we could get it merged soon.

bondjimbond commented:
@mghughes If we can get our Travis issues settled it should be clear for merging. If it needs to get merged sooner than that can happen, I wouldn't mind getting a second pair of eyes on it just for coding standards. But it looks fine to me.

bondjimbond commented:
I don't want progress on Islandora to halt because Travis is broken, so Merge it is! Hopefully it doesn't come up with any complaints when it's back online, but the code looks OK to me.
